### ISSUE TITLE
Sending packet exceeds MTU due to Ethernet layer

### ISSUE THREAD

**Issue by The7Tycoon** [seladb/PcapPlusPlus#598]
I have a PCAP containing UDP packets that have been fragmented due to being well over the MTU. The first fragments fully use the MTU, meaning the IP layer is 1500 bytes long (IP header + payload). Then, when they were sent, an traditional 14 bytes Ethernet layer was added, so the packet is 1514 bytes long in total.

However, now that I want to send them back over the network, I'm having trouble with PcapLiveDevice::sendPacket(). It takes the whole raw packet, including the 14 bytes link layer, and determines that the packet is bigger than the 1500 bytes MTU, when I believe the Ethernet header should not be taken into consideration. 

I tried sending only the data starting at the IP layer (so exactly 1500 bytes), no error is shown but I can't see any packets on wireshark. The only temporary solution was to increase the MTU to 1514 so that the checks in sendPacket() passes.

Am I doing something wrong ? Or should sendPacket() only count the data length at the IP layer ?

All of this is tried on Linux (CentOS 8.2), with a default MTU of 1500.

**Comment by seladb** [comment]
hi @The7Tycoon I'm not sure I fully understand your scenario: the first fragment is 1500 bytes long but doesn't contain the Ethernet header? How did you capture it then? Which interface are you using? 
And regarding sending these packets: how were these 14 bytes of the Ethernet header added? Did you add them?
Basically, you can't send packets that exceed the MTU but I'm not sure which interfaces are used for capturing and sending packets so it's hard to pinpoint the issue

**Comment by SamuelBailey** [comment]
# Problem
I have come across this exact issue. I captured UDP packets from a PcapLiveDevice, and then attempted to forward the packet using PcapLiveDevice::sendPacket(). This worked fine for small packets, however the size check for the MTU (1500 bytes) meant that the device refused to send for larger packets, because it was recording a size of 1514 bytes (including the ethernet header). Looking this up, the MTU should only include the IP layer size, not the size of the Data Link Layer header, so the check should ether:
1. Compare length of the **packet** (**including data link layer**) against (**MTU** + **Data Link layer**)

or

2.  Find the size of the **IP layer** contained within the packet and compare this against **MTU**.

I have my own public fork of PcapPlusPlus in which I patched this for use in my own project. I hadn't opened a pull request because when I cloned the repository and just ran the tests (before making any changes), I couldn't get all of them to pass. 

I can, however, point you to the exact location where this occurs and explain how I fixed it. (I used the second method stated above).

# Where this occurs

Within `PcapPlusPlus/Pcap++/src/PcapLiveDevice.cpp`, I was calling the function
```c++
bool PcapLiveDevice::sendPacket(RawPacket const& rawPacket)
```
which in turn calls
```c++
bool PcapLiveDevice::sendPacket(const uint8_t* packetData, int packetDataLength)
```
with `packetDataLength` set to the length of the entire packet, including the Ethernet header.

The 3rd `if` statement inside `bool PcapLiveDevice::sendPacket(const uint8_t* packetData, int packetDataLength)` as shown below, checks `packetDataLength` against `m_DeviceMtu`:
```c++
if (packetDataLength > (int)m_DeviceMtu)
{
	LOG_ERROR("Packet length [%d] is larger than device MTU [%d]\n", packetDataLength, (int)m_DeviceMtu);
	return false;
}
```
An error is thrown if this check does not pass. If you disable this check then PcapPlusPlus will successfully send packets with an IP layer size of 1500 bytes, so Data Link header + IP = 1514 bytes (if using ethernet).

# My patch
The way I fixed this was to partially parse the `RawPacket` before it was passed to the `bool PcapLiveDevice::sendPacket(const uint8_t* packetData, int packetDataLength)` function.

In my own solution, I added an additional field called `int packetPayloadLength` which did not include the Ethernet header size to the sendPacket function, so its new signature was this:
```c++
bool PcapLiveDevice::sendPacket(const uint8_t* packetData, int packetDataLength, int packetPayloadLength)
```
The changes I made are available on [my fork of PcapPlusPlus](https://github.com/SamuelBailey/PcapPlusPlus/tree/packet-size) in the branch named **packet-size**.

The commit for this fix is (currently) the most recent commit in the **packet-size** branch:  
[07e74d36bdf6d98eda3c45b9c5a7071cd7e9ec62](https://github.com/SamuelBailey/PcapPlusPlus/commit/07e74d36bdf6d98eda3c45b9c5a7071cd7e9ec62)

**Comment by seladb** [comment]
Thank you so much @SamuelBailey for the thorough analysis and explanation!
This is definitely a bug exists for a long time...

I looked at your solution and it seems right to me, would you consider opening a PR?
Regarding running the tests - maybe this might help: https://pcapplusplus.github.io/docs/tests

One concern that I have is around the performance penalty we might see from parsing each and every packet (even though we're only parsing up to the network layer). One solution I can suggest is to let the user decide whether to perform this check or not. In certain use-cases the user might not want to have this penalty if they're certain packets won't go over the MTU. Please let me know what you think

**Comment by SamuelBailey** [comment]
Thanks, @seladb. I'll have another look at running the tests soon and will let you know how that goes. If I can get it running I'll see if I can add a test for the MTU and create a pull request.

I think it's sensible to give the user the option to disable the check as you mentioned, given the overhead of partially parsing the packet. I can quite simply add a parameter `bool checkMtu` with a default value of `true` and then perform the check conditionally on this.

**Comment by SamuelBailey** [comment]
I have tried running the tests again. I checked the most up-to-date version of your repo, and have tried testing in both the **master** and **dev** branches (without any changes just to see if I could get the tests to pass). My system is Ubuntu 20.04.
I configured my system by running 
```
./configure-linux.sh --default
```
and then compiling and installing.  
As far as I can tell, my system is set up the same as the Ubuntu 20.04 machine in your CI.

I noticed that the CI only runs tests which don't require a network interface, using the command
```
cd Tests/Pcap++Test
Bin/Pcap++Test -n
```
When I perform this command the tests all pass, however if I use my network interface, as shown here:
```
cd Tests/Pcap++Test
sudo Bin/Pcap++Test -i 192.168.1.122
```
they will pass up to, and including TestWinPcapLiveDevice, at which point the test suite hangs and doesn't finish (I've tried leaving it for several minutes). I found that restarting my network interface card at this point causes the remaining tests to complete, however the following tests, which require the network interface to be up, will then fail.
* TestSendPacket
* TestSendPackets
* TestPcapFiltersLive
* TestRawSockets

Since TestSendPacket is directly after TestWinPacpLiveDevice, I believe this is the test that is not finishing. I haven't yet looked in the test suite itself to see what's going wrong exactly. In the next day or two I should be able to take a look. I will likely need to take a look at this test anyway, if I am to add the checks for MTU.

If there's anything I'm missing here, please let me know.

**Comment by seladb** [comment]
Thanks for the update @The7Tycoon ! I think your analysis is accurate - the issue is probably in `TestSendPacket`. It's hard to guess why it hangs, I may be able to assist if you can debug it a little more and provide more information on where exactly you think it hangs.

**Comment by SamuelBailey** [comment]
I looked into the testing. I think there may be a bug in my network card driver on my desktop. I added logging to `TestSendPacket` (inside `PcapPlusPlus/Tests/Pcap++Test/LiveDeviceTests.cpp`). This test sends a large number of packets from `PcapExamples/example.pcap` over the interface. I found that if I restarted my network card between runs, it would hang when attempting to send the 67th packet over the `liveDev`. If I did **not** restart my network card between runs, it would hang when attempting to send the 53rd packet over the `liveDev`.

I then tested on my laptop to see if it was something to do with my configuration, and all tests passed without a problem.

**TLDR**: I think my network driver on my desktop contains a bug. I should be able to test on my laptop in the meantime and hopefully complete the pull request soon.

**Comment by seladb** [comment]
Thanks for the info @SamuelBailey !
I'm curious about which part of the code it hangs. Is it inside the PcapPlusPlus code or WinPcap/libpcap?

**Comment by SamuelBailey** [comment]
I've created the changes mentioned earlier and created a unit test for it. I merged my changes into the **dev** branch of my own fork. Following your guides on how to create a pull request, I've created accounts in Appveyor and TravisCI, however I'm not familiar with these systems. What do I need to do to run the tests on the dev branch in my fork?

Also, I just want to confirm a couple of things were done the way you'd want:  

I added a couple of parameters to the `sendPacket` methods to determine whether you want to perform the MTU check. If you are passing the packet as a raw byte array, I set the default value for the `checkMtu` to **false**, since parsing the packet requires knowing which Data Link Layer header you are using (this can be specified by another parameter which has a default value of `pcpp::LINKTYPE_ETHERNET`), so this should keep it backwards compatible with everything from before. If you pass the packet in as a `Packet` or `RawPacket` object, the MTU check is **true** by default, since the Data Link type is known.

Also, I added a test inside `LiveDeviceTests.cpp` for the MTU. It uses some of the `layer` libraries to construct packets, however this did mean adding a few additional header files to the `LiveDeviceTests.cpp` file.

> I'm curious about which part of the code it hangs. Is it inside the PcapPlusPlus code or WinPcap/libpcap?
I haven't had time to trace that just yet. I will probably take a look into that tomorrow or this weekend, and let you know if I manage to find where exactly the problem lies.

**Comment by seladb** [comment]
First and foremost, I wanted to thank you again for all your dedication and hard work, I really appreciate it!
Regarding your questions:

> I've created the changes mentioned earlier and created a unit test for it. I merged my changes into the **dev** branch of my own fork. Following your guides on how to create a pull request, I've created accounts in Appveyor and TravisCI, however I'm not familiar with these systems. What do I need to do to run the tests on the dev branch in my fork?

Appveyor, TravisCI, GitHub Actions and CirrusCI are all CI platforms used in PcapPlusPlus. The reason I use all of them is because each provides a slightly different set of capabilities, especially around OS and platform support. Actually the `CONTRIBUTING.md` file is out of date because today most of the CI load moved to GitHub Actions and CirrusCI. Anywat, in order to run the CI the best option would probably be to create a PR - once a PR is created it will run in PcapPlusPlus main repo which is visible for everyone.
You can create your PR as a "draft" until you're sure with your code and then move it to "ready to review". That way I'll know when to review it.

> I added a couple of parameters to the `sendPacket` methods to determine whether you want to perform the MTU check. If you are passing the packet as a raw byte array, I set the default value for the `checkMtu` to **false**, since parsing the packet requires knowing which Data Link Layer header you are using (this can be specified by another parameter which has a default value of `pcpp::LINKTYPE_ETHERNET`), so this should keep it backwards compatible with everything from before. If you pass the packet in as a `Packet` or `RawPacket` object, the MTU check is **true** by default, since the Data Link type is known.

I mostly agree with everything you said, my only comment would be: for `RawPacket` I'd suggest keeping `checkMtu` default value as `false` because `true` would require parsing the packet which may have (small) performance penalty. Keeping it as `false` will maintain full backward compatibility.

> Also, I added a test inside `LiveDeviceTests.cpp` for the MTU. It uses some of the `layer` libraries to construct packets, however this did mean adding a few additional header files to the `LiveDeviceTests.cpp` file.

That should be fine, I don't see any issue with adding additional `#include` to `LiveDeviceTests.cpp`

**Comment by SamuelBailey** [comment]
No worries, @seladb! Your quick responses have been very useful in getting things done the right way. Apologies I took a while to get back to this issue. I moved things around to ensure that I'm not parsing any packets unnecessarily and am currently running the test suite. I also realised I need to add the `checkMtu` flag to `sendPackets` as well as `sendPacket` methods. I intend to get this finished and a pull request up sometime tomorrow.

**Comment by seladb** [comment]
thank you @SamuelBailey , I really appreciate your efforts on this issue! Looking forward to your PR

**Comment by SamuelBailey** [comment]
I've created the PR and fixed an error which was flagged by the CI. I don't think I am able to link it to this issue - It is PR #614

**Comment by seladb** [comment]
Thank you so much @SamuelBailey ! I'll go over the PR in the coming days

**Comment by SamuelBailey** [comment]
Thanks very much! If there's anything where it's unclear about what I've done or why, then just let me know, and I'll try to get back to you as soon as possible

**Comment by The7Tycoon** [comment]
I've not been very active since I created the issue but I want to thank you both for taking the time to look into it and fix it !

**Comment by seladb** [comment]
Thanks @The7Tycoon ! The real person you should be thanking is @SamuelBailey who is making great efforts to fix this!
Feel free to look at the PR https://github.com/seladb/PcapPlusPlus/pull/614 and make your comments

**Comment by SamuelBailey** [comment]
I've made the changed mentioned in the PR #614 , however when pushing my changes, the CI was failing because a docker pull was unsuccessful

**Comment by SamuelBailey** [comment]
PR #614 is open for review again. I retriggered the checks this time there weren't any issues with the docker pull in the CI.

**Comment by seladb** [comment]
@SamuelBailey thank you so much for your contribution!!

Your PR is merged to `master` so I think we can close this issue.
